nucypher/blockchain/eth/deployers.py

```diff
@@ -222,72 +231,125 @@ def deploy(self, secret_hash: bytes, gas_limit: int, progress) -> dict:
 
-        # 3 - Wrap the escrow contract #
+        # 4 - Wrap the escrow contract
         wrapped_contract = self.blockchain._wrap_contract(wrapper_contract=proxy_deployer.contract,
                                                           target_contract=new_contract)
 
-        # 4 - Set the new Dispatcher target #
+        # 5 - Set the new Dispatcher target
         upgrade_receipt = proxy_deployer.retarget(new_target=new_contract.address,
                                                   existing_secret_plaintext=existing_secret_plaintext,
                                                   new_secret_hash=new_secret_hash,
                                                   gas_limit=gas_limit)
 
-        # 5- Respond
+        # 6 - Respond
         upgrade_transaction = {'deploy': deploy_receipt, 'retarget': upgrade_receipt}
         self._contract = wrapped_contract  # Switch the contract for the wrapped one
         return upgrade_transaction
 
     def rollback(self, existing_secret_plaintext: bytes, new_secret_hash: bytes, gas_limit: int = None):
+        """
+        Execute an existing deployments proxy contract, engaging the upgrade rollback intrfaces,
+        modifying the proxy's on-chain contract target to the most recet previous target.
+        """
+
         if not self._upgradeable:
             raise self.ContractNotUpgradeable(f"{self.contract_name} is not upgradeable.")
 
         existing_bare_contract = self.blockchain.get_contract_by_name(registry=self.registry,
                                                                       name=self.contract_name,
                                                                       proxy_name=self._proxy_deployer.contract_name,
                                                                       use_proxy_address=False)
-        proxy_deployer = self._proxy_deployer(registry=self.registry,
-                                              target_contract=existing_bare_contract,
-                                              deployer_address=self.deployer_address,
-                                              bare=True)  # acquire agency for the proxy itself.
-
+        proxy_deployer = self.get_proxy_deployer(registry=self.registry, provider_uri=self.blockchain.provider_uri)
         rollback_receipt = proxy_deployer.rollback(existing_secret_plaintext=existing_secret_plaintext,
                                                    new_secret_hash=new_secret_hash,
                                                    gas_limit=gas_limit)
 
         return rollback_receipt
 
+    def _finish_deploy_essential(self, deployment_receipt: dict, progress = None):
+        """Used to divert flow control for bare contract deployments."""
+        deployment_step_name = self.deployment_steps[0]
+        result = {deployment_step_name: deployment_receipt}
+        if progress:
+            progress.update(len(deployment_step_name))  # Update the progress bar to completion.
```

Comment:
I'm not sure I follow - the length of the step name?

Reply:
I am trying to access the *number of deployment steps* - is there a better way?

Reply:
oh did you mean to use:
```suggestion
            progress.update(len(deployment_steps))  # Update the progress bar to completion.
```

I was confused because you were using the lenght of the specific step name, not the lenght of the list of steps.

Reply:
I do feel this is a bit awkward, but sometimes, we just want to perform part of the deployment series, without all the steps included.  This dehydrates shared logic for that diversion.  Perhaps there is a better way?

Reply:
Ok, now I get it :)
My only comment now to this respect is that in the `initial_deployment` case, we store all deployment receipts in `self.deployment_receipts`, and this is missing here. 
BTW, I also realized we don't do that in all deployers, so perhaps this is a good moment for either do it in all of them or in none.

Reply:
:heavy_check_mark: